keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,877 @@
+    // resources created from this Notification.
+    // +optional
+    SucceededCount int
+
+    // If set, it suggests the timestamp(UTC) at when the first time the
+    // Notification has been acknowledged by the Notification Controller.
+    // +optional
+    StartTime *metav1.Time
+
+    // If set, it suggests the timestamp(UTC) at when the Notification has
+    // been marked as completed. A Notification is completed when all
+    // PodNotifications created from this Notification have completed.
+    // Note that this field will never be set if "AllPods" is specified as
+    // pod selection policy.
+    // +optional
+    CompleteTime *metav1.Time
+}
+```
+
+##### ContainerNotifierHandler
+```
+// ContainerNotifierHandler defines a specific action to be executed or a signal which will be delivered.
+// Exactly one of Exec and Signal should be set.
+type ContainerNotifierHandler struct {
+    // Exec specifies the action to take.
+    // +optional
+    Exec *ExecAction `json:"exec,omitempty" protobuf:"bytes,1,opt,name=exec"`
+
+    // Signal specifies a signal to send to the container
+    // +optional
+    // define constants for signals?
+    // validate the signals are valid?  windows?
+++    Signal string
+}
+```
+
+#### Phase 3 API Additions
+
+If needed, we may add a probe to ContainerNotifier in Phase 3 to verify that the ContainerNotifierHandler has delivered the signal or executed the command.
+
+Phase 3 is more loosely decided than phases 1 and 2 at this stage, we will have to revisit the design details.
+
+```
+type ContainerNotifier struct {
+    // Name of the ContainerNotifier specified as a DNS_LABEL. Each
+    // ContainerNotifier within a Container must have a unique name.
+    // Names with the prefix `k8s.io/` are reserved for Kubernetes-defined
+    // "well-known" notifiers.
+    // Names with other prefixes such as `example.com/` are custom names.
+    // Immutable.
+    Name string
+
+    // The action to take when a matching PodNotification is observed.
+    // Exactly-once execution of the action is NOT guaranteed, so it's user's
+    // responsibility to make sure the handler is idempotent to get consistent
+    // behavior.
+    // For example, when a quiesce command has been ran on a database, the
+    // database will stay quiesced, and any consecutive execution of the
+    // same quiesce command should not have additional impact.
+    Handler *ContainerNotifierHandler
+
+    // Number of seconds after which the notifier will time out.
+    // Default to 1 second. Minimum value is 1.
+    // +optional
+    TimeoutSeconds int32
+
++   // Add a Probe in Phase 3. Probe defined in the core API here will be used.
++   Probe *Probe
+}
+```
+
+### Kubelet Impact in Phase 2 and Beyond
+
+When moving the logic from the PodNotification controller to Kubelet, we foresee the following changes to happen in Kubelet:
+
+- Kubelet will watch PodNotification resources cluster-wide.
```

Comment:
It only needs to watch notifications for pods on this node.  That's a potentially important optimization we should plan for,

Reply:
agreed. Updated this point.